File: LinkedList/622. Design Circular Queue/Solution.java
class ListNode {
    public int val;
    public ListNode next;

    ListNode() {
        this.val = 0;
        this.next = null;
    }

    ListNode(int val) {
        this.val = val;
        this.next = null;
    }

    ListNode(int val, ListNode node) {
        this.val = val;
        this.next = node;
    }
}

class MyCircularQueue {
    // N : total no of nodes
    // Time Complexity: O(1)
    // Space Complexity: O(1)
    private static int MAX_LEN;
    private ListNode front;
    private ListNode rear;
    private int length;

    public MyCircularQueue(int k) {
        MAX_LEN = k;
        front = null;
        rear = null;
        length = 0;
    }

    public boolean enQueue(int value) {
        if (isFull())
            return false;

        ListNode new_node = new ListNode(value);
        if (isEmpty())
            front = rear = new_node;
        else {
            rear.next = new_node;
            rear = rear.next;
            new_node.next = front;
        }
        length++;
        return true;
    }

    public boolean deQueue() {
        if (isEmpty())
            return false;

        front = front.next;
        rear.next = front;
        length--;
        return true;
    }

    public int Front() {
        return (isEmpty()) ? -1 : front.val;
    }

    public int Rear() {
        return (isEmpty()) ? -1 : rear.val;
    }

    public boolean isEmpty() {
        return (length == 0) ? true : false;
    }

    public boolean isFull() {
        return (length == MAX_LEN) ? true : false;
    }
}

File: LinkedList/622. Design Circular Queue/Solution.py
class ListNode:
    def __init__(self, value = 0, node = None):
        self.val = value
        self.next = node

class MyCircularQueue:
    # N : total no of nodes
    # Time Complexity: O(1)
    # Space Complexity: O(1)
    def __init__(self, k: int):
        self.MAX_LEN = k
        self.length = 0
        self.front = None
        self.rear = None

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False

        new_node = ListNode(value, None)
        if self.isEmpty():
            self.front = self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = self.rear.next
            self.rear.next = self.front
        self.length += 1
        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False
        
        self.front = self.front.next
        self.rear.next = self.front
        self.length -= 1
        return True

    def Front(self) -> int:
        if self.isEmpty():
            return -1
        return self.front.val

    def Rear(self) -> int:
        if self.isEmpty():
            return -1
        return self.rear.val

    def isEmpty(self) -> bool:
        return True if self.length == 0 else False

    def isFull(self) -> bool:
        return True if self.length == self.MAX_LEN else False